Ndev 3380  compasability for iterative trx

kristinaNikolaevaa opened the pull request.

kristinaNikolaevaa committed:
added iterative tests for compasibility

kristinaNikolaevaa committed:
added proxy tests

kristinaNikolaevaa committed:
Merge branch 'develop' into NDEV-3380

kristinaNikolaevaa committed:
added one more test for evm

kristinaNikolaevaa committed:
Merge branch 'develop' into NDEV-3380

romanova-natasha committed:
fix test_step_from_instruction_for_counter: use holder account with function scope

fix test_step_from_instruction_for_counter: use holder account with function scope

romanova-natasha committed:
fix iterative transaction tests: reduce iteration lenght

fix iterative transaction tests: reduce iteration lenght

kristinaNikolaevaa committed:
Merge branch 'develop' into NDEV-3380

romanova-natasha committed:
fix: new holder account for all interoperability tests

fix: new holder account for all interoperability tests

romanova-natasha committed:
add composability test cases: iterative actions + solana call/calls, deploy contracy and solana call

add composability test cases: iterative actions + solana call/calls, deploy contracy and solana call

romanova-natasha committed:
fix deploy contract and solana call test

fix deploy contract and solana call test

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -303,3 +304,185 @@ def test_limit_of_simple_instr_in_one_trx(self, call_solana_caller, counter_reso
         resp = self.web3_client.send_transaction(sender, instruction_tx)
         assert resp["status"] == 0
+
+    @pytest.mark.parametrize("iterations_number", [29, 52])
```
one value is enough

romanova-natasha replied:
done

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -303,3 +304,185 @@ def test_limit_of_simple_instr_in_one_trx(self, call_solana_caller, counter_reso
+            call_solana_caller.functions.executeInIterativeMode(iterations, lamports, serialized).build_transaction(tx)
+
+    @pytest.mark.parametrize("iterations_number", [40, 55])
+    def test_iterative_actions_no_solana_call(self, iterations_number, call_solana_caller):
```
why do you need this test without solana call?

romanova-natasha replied:
removed

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -303,3 +304,185 @@ def test_limit_of_simple_instr_in_one_trx(self, call_solana_caller, counter_reso
+        event_logs = call_solana_caller.events.LogBytes().process_receipt(resp)
+        assert int.from_bytes(event_logs[0].args.value, byteorder="little") == next(get_counter_value)
+
+    def test_iterative_actions_after_solana_call_exceed_accounts_limit(self, counter_resource_address, call_solana_caller, get_counter_value):
```
this test can be deleted 
iterative actions after solana call can't be called at all

romanova-natasha replied:
It is working now and idk why, I asked the core team about this and wait for their response

kristinaNikolaevaa commented on contracts/precompiled/CallSolanaCaller.sol:
```diff
@@ -28,9 +34,48 @@ contract CallSolanaCaller {
+    function deployStorageAndCallSolana(string memory message, uint64 lamports, bytes calldata instruction) public {
+        Storage storageContract = new Storage();
+        execute(lamports, instruction);
+        emit LogStr(message);
```
it is better to return storageContract address and try to call it in the test or at least to check that address exist and contains some code

romanova-natasha replied:
I added call contract function and I return contract address

kristinaNikolaevaa commented on contracts/precompiled/CallSolanaCaller.sol:
```diff
@@ -28,9 +34,48 @@ contract CallSolanaCaller {
+        doIterativeActions(iterations);
+    }
+
+    function executeMultipleCallsAndDoIterativeActions(uint256 calls, uint256 iterations, uint64 lamports, bytes calldata instruction) public {
```
can we rename iterations to something else? because iterations is a term from evm and it is a bit confusing
this iterations count != real iterations for the transaction

romanova-natasha replied:
done

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -303,3 +304,185 @@ def test_limit_of_simple_instr_in_one_trx(self, call_solana_caller, counter_reso
+        event_logs = call_solana_caller.events.LogBytes().process_receipt(resp)
+        assert int.from_bytes(event_logs[0].args.value, byteorder="little") == next(get_counter_value)
+    
+    def test_multiple_solana_calls_and_iterative_actions_instructions_limit(self, counter_resource_address, call_solana_caller, get_counter_value):
```
I think we don't need this test
the same as with 1 call

romanova-natasha replied:
There was an issue, where the instruction limit test had failed for an iterative transaction only. I kept this test to check the issue is fixed, and I still want to keep it. What do you think?

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -303,3 +304,185 @@ def test_limit_of_simple_instr_in_one_trx(self, call_solana_caller, counter_reso
+        resp = self.web3_client.send_transaction(sender, instruction_tx)
+        assert resp["status"] == 1
+        event_logs = call_solana_caller.events.LogBytes().process_receipt(resp)
+        assert int.from_bytes(event_logs[0].args.value, byteorder="little") == next(get_counter_value)
```
what the value of next(get_counter_value) here? I think it should be = solana_calls

romanova-natasha replied:
fixed: I check a number of logged solana calls now

romanova-natasha committed:
rm tests with no iterative txs check

rm tests with no iterative txs check

romanova-natasha committed:
check holder account state after tx execution

check holder account state after tx execution

romanova-natasha committed:
rename test test_solana_call_before_iterative_actions

rename test test_solana_call_before_iterative_actions

romanova-natasha committed:
add case: eip 1559 iterative tx check with program_id = transfer_tokens_id

add case: eip 1559 iterative tx check with program_id = transfer_tokens_id

romanova-natasha committed:
add case: failed solana call

add case: failed solana call

romanova-natasha committed:
add case: iterative tx with solana call in sol network

add case: iterative tx with solana call in sol network

romanova-natasha committed:
rm dbg info

rm dbg info

romanova-natasha committed:
add case: iterative tx with send tokens

add case: iterative tx with send tokens

romanova-natasha committed:
add case of solana call with 2 programs, minor refactoring

add case of solana call with 2 programs, minor refactoring

romanova-natasha committed:
fix review

fix review

artem-yazkov committed:
NDEV-3380 change tests order for tests test_iterative_actions_and_multiple_solana_calls_instructions_limit, test_transfer_with_pda_signature_iterative_tx_eip_1559

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -66,6 +123,58 @@ def test_counter_execute_with_get_return_data(
+    def test_transfer_with_pda_signature_iterative_tx_eip_1559(self, call_solana_caller, sol_client, solana_account, pytestconfig, bank_account):
+        iterations = 20
+        sender = self.accounts[0]
+        from_wallet = Keypair()
```
you can you exist fixture solana_account (solana account with sol)

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -303,3 +412,251 @@ def test_limit_of_simple_instr_in_one_trx(self, call_solana_caller, counter_reso
+                                                                                       lamports,
+                                                                                       serialized).build_transaction(tx)
+        resp = self.web3_client.send_transaction(sender, instruction_tx)
+        assert resp["status"] == 1
```
It shouldn't work
please ask core team why it works now

romanova-natasha replied:
it should work like this, I also added a negative scenario, where an evm memory limit exceeded

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -303,3 +412,251 @@ def test_limit_of_simple_instr_in_one_trx(self, call_solana_caller, counter_reso
+        event_logs = call_solana_caller.events.LogBytes().process_receipt(resp)
+        assert int.from_bytes(event_logs[0].args.value, byteorder="little") == next(get_counter_value)
+
+    def test_iterative_actions_after_solana_call_exceed_accounts_limit(self, counter_resource_address, call_solana_caller):
```
I think  this test can be deleted as previous one is negative

romanova-natasha replied:
I want to keep it cause the reason here is account limit, previous test is positive, the third one is negative but with evm memory limit

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -303,3 +412,251 @@ def test_limit_of_simple_instr_in_one_trx(self, call_solana_caller, counter_reso
+                                                                                     serialized).build_transaction(tx)
+        resp = self.web3_client.send_transaction(sender, instruction_tx)
+        assert resp["status"] == 1
+        assert len(call_solana_caller.events.LogBytes().process_receipt(resp)) == solana_calls
```
does it work when you run all tests?
you use counter_resource_address for several tests and it means  len(call_solana_caller.events.LogBytes().process_receipt(resp)) == solana_calls + N

N - previous runs of COUNTER_ID program

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -303,3 +412,251 @@ def test_limit_of_simple_instr_in_one_trx(self, call_solana_caller, counter_reso
+        serialized = serialize_instruction(COUNTER_ID, instruction)
+
+        tx = self.web3_client.make_raw_tx(sender.address)
+        instruction_tx = call_solana_caller.functions.deployStorageAndCallSolana("deploy contracts status: done",
```
why do you need to send there such string

romanova-natasha replied:
I don't, removed it. (commited it by accident)

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -303,3 +412,251 @@ def test_limit_of_simple_instr_in_one_trx(self, call_solana_caller, counter_reso
+        assert resp["status"] == 1
+
+        event_logs = call_solana_caller.events.LogAddress().process_receipt(resp)
+        assert event_logs[0].args.value is not None
```
you can check here also counter value

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -303,3 +412,251 @@ def test_limit_of_simple_instr_in_one_trx(self, call_solana_caller, counter_reso
+        assert resp["status"] == 1
+
+        balance_after = self.web3_client.get_balance(call_solana_caller.address)
+        assert balance_after == balance_before + 10
```
you can check here also counter value

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -303,3 +412,251 @@ def test_limit_of_simple_instr_in_one_trx(self, call_solana_caller, counter_reso
+    def test_solana_call_of_two_programs_in_one_iterative_tx(self, counter_resource_address, call_solana_caller, sol_client, solana_account, pytestconfig, bank_account):
+        iterations = 10
+        sender = self.accounts[0]
+        from_wallet = Keypair()
```
please use exist solana_account fixture

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -303,3 +412,251 @@ def test_limit_of_simple_instr_in_one_trx(self, call_solana_caller, counter_reso
+        else:
+            sol_client.request_airdrop(from_wallet.pubkey(), 1000 * 10**9, commitment=Confirmed)
+
+        serialized, mint, accounts_list = self.get_transfer_instruction(sol_client, from_wallet, to_wallet, amount, call_solana_caller)
```
serialized_transfer?

romanova-natasha replied:
done

kristinaNikolaevaa replied:
looks the same

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -303,3 +412,251 @@ def test_limit_of_simple_instr_in_one_trx(self, call_solana_caller, counter_reso
+        assert resp["status"] == 1
+        assert int(mint.get_balance(accounts_list[1], commitment=Confirmed).value.amount) == amount
+
+        wait_condition(
```
please check counter value

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -41,6 +43,61 @@ class TestSolanaInteroperability:
     web3_client: NeonChainWeb3Client
 
+    @pytest.fixture(scope="class")
+    def account_sol_network(self, class_account_sol_chain):
```
why do you need the fixture which return other fixture without any changes

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on integration/tests/basic/evm/test_solana_interoperability.py:
```diff
@@ -41,6 +43,61 @@ class TestSolanaInteroperability:
+            ],
+            data=bytes([0x0]),
+        )
+        return serialize_instruction(TRANSFER_TOKENS_ID, instruction), mint, [from_token_account, to_token_account]
```
It is very difficult to read when some function return more then 1 value
You can leave it as it is if it is the only option :)

romanova-natasha committed:
changes after the review

changes after the review

romanova-natasha committed:
add matrix elems sum calculation combined with solana calls

add matrix elems sum calculation combined with solana calls

romanova-natasha committed:
add solana call inside matrix elems sum calculation case

add solana call inside matrix elems sum calculation case

romanova-natasha committed:
rm useless case

rm useless case

romanova-natasha committed:
Merge remote-tracking branch 'origin/develop' into NDEV-3380

romanova-natasha committed:
check counter value

check counter value

romanova-natasha committed:
fix counter value for test checks, check logs in tests with 2 different instructions in one tx

fix counter value for test checks, check logs in tests with 2 different instructions in one tx

romanova-natasha committed:
counter_resource_address fixture change: get a unique resource for each test

counter_resource_address fixture change: get a unique resource for each test

romanova-natasha committed:
set scope class for resource address and local counter, fix test with multiple solana calls

set scope class for resource address and local counter, fix test with multiple solana calls

romanova-natasha committed:
Merge remote-tracking branch 'origin/develop' into NDEV-3380

# Conflicts:
#	integration/tests/neon_evm/test_interoperability.py
#	integration/tests/neon_evm/utils/call_solana.py

romanova-natasha committed:
change fixtures scope for interoperability tests

change fixtures scope for interoperability tests

romanova-natasha committed:
add composability case: check 2 txs parallel execution from one sender

add composability case: check 2 txs parallel execution from one sender

romanova-natasha committed:
Merge remote-tracking branch 'origin/develop' into NDEV-3380

# Conflicts:
#	integration/tests/neon_evm/test_interoperability.py

romanova-natasha committed:
fix conflicts after merge

fix conflicts after merge

kristinaNikolaevaa commented on integration/tests/neon_evm/test_interoperability.py:
```diff
@@ -446,3 +450,80 @@ def test_call_neon_instruction_by_neon_instruction(
+        resp = solana_caller.execute(program_id=COUNTER_ID, instruction=instruction, sender=sender_with_tokens, holder_acc=new_holder_acc)
+        check_transaction_logs_have_text(solana_client, trx=resp, text="exit_status=0x11")
+
+        operator_balance_pubkey = evm_loader.get_operator_balance_pubkey(operator_keypair)
```
you already have operator_balance_pubkey

kristinaNikolaevaa commented on contracts/precompiled/CallSolanaCaller.sol:
```diff
@@ -17,73 +23,200 @@ contract CallSolanaCaller {
+        emit LogAddress(address(storageContract));
+    }
+
+    function execute_with_get_return_data(
```
let's make all functions name in one style

romanova-natasha committed:
fix test_revision_2_txs_from_1_sender: add shared context within contract

fix test_revision_2_txs_from_1_sender: add shared context within contract

romanova-natasha committed:
add composability case: balance account revision changed

add composability case: balance account revision changed

romanova-natasha committed:
minor refactoring

minor refactoring

romanova-natasha committed:
Merge remote-tracking branch 'origin/develop' into NDEV-3380

romanova-natasha committed:
fix lint

fix lint

romanova-natasha committed:
isolate composability revision check tests with unique holder accounts

isolate composability revision check tests with unique holder accounts

kristinaNikolaevaa committed:
fixed all tests

kristinaNikolaevaa committed:
removed unused function

kristinaNikolaevaa committed:
Update test_rpc_estimate_gas.py

kristinaNikolaevaa committed:
Update test_rpc_estimate_gas.py

kristinaNikolaevaa committed:
Merge branch 'develop' into NDEV-3380
